Firestore: Fix sorting 'delete_changes' in 'Watch._compute_snapshot'.

liamuk opened the pull request:
`delete_changes` on the changed line is a list of string, not a list of DocumentSnapshots, so it should not be sorted with the query comparator function (see https://github.com/googleapis/google-cloud-python/blob/master/firestore/google/cloud/firestore_v1/watch.py#L575)

For context, I'm getting this error in one of my snapshot listeners in google-cloud-firestore==1.3.0
```
Thread-ConsumeBidirectionalStream caught unexpected exception 'str' object has no attribute 'reference' and will exit.
Traceback (most recent call last):
  File ".../venv/lib/python3.5/site-packages/google/api_core/bidi.py", line 635, in _thread_main
    self._on_response(response)
  File ".../venv/lib/python3.5/site-packages/google/cloud/firestore_v1/watch.py", line 443, in on
_snapshot
    meth(proto)
  File ".../venv/lib/python3.5/site-packages/google/cloud/firestore_v1/watch.py", line 378, in _o
n_snapshot_target_change_no_change
    self.push(change.read_time, change.resume_token)
  File ".../venv/lib/python3.5/site-packages/google/cloud/firestore_v1/watch.py", line 540, in pu
sh
    self.doc_tree, self.doc_map, deletes, adds, updates
  File ".../venv/lib/python3.5/site-packages/google/cloud/firestore_v1/watch.py", line 669, in _c
ompute_snapshot
    delete_changes = sorted(delete_changes, key=key)
  File ".../venv/lib/python3.5/site-packages/google/cloud/firestore_v1/query.py", line 819, in _c
omparator
    comp = Order()._compare_to(doc1.reference._path, doc2.reference._path)
AttributeError: 'str' object has no attribute 'reference'
```

liamuk commented:
@googlebot I signed it!

tseaver commented:
@liamuk Thanks for the patch!  Our tests are passing here because they do not exercise `Watch._compute_snapshot` on an instance with a non-dummy comparator.  Can you please add a test to the `TestWatch` class `tests/unit/v1/test_watch.py` which exercises that case?  E.g.:

```python
    def test__compute_snapshot_deletes_w_real_comparator(self):
        from google.cloud.firestore_v1.watch import WatchDocTree

        doc_tree = WatchDocTree()

        class DummyDoc(object):
            update_time = mock.sentinel

        def _comparator(doc1, doc2):
            if doc1.field > doc2.field:
                return 1
            if doc1.field < doc2.field:
                return -1
            return 0

        deleted_doc_1 = DummyDoc()
        deleted_doc_2 = DummyDoc()
        doc_tree = doc_tree.insert(deleted_doc_1, None)
        doc_tree = doc_tree.insert(deleted_doc_2, None)
        doc_map = {
            "/deleted_1": deleted_doc_1,
            "/deleted_2": deleted_doc_2,
        }
        delete_changes = ["/deleted_1", "/deleted_2"]
        add_changes = []
        update_changes = []
        inst = self._makeOne(comparator=_comparator)
        updated_tree, updated_map, applied_changes = inst._compute_snapshot(
            doc_tree, doc_map, delete_changes, add_changes, update_changes
        )
        self.assertEqual(updated_map, {})
```

liamuk commented:
Added that test verbatim

tseaver commented on firestore/tests/unit/v1/test_watch.py:
```diff
@@ -735,6 +735,38 @@ class DummyDoc(object):
+                return 1
+            if doc1.field < doc2.field:
+                return -1
+            return 0
```
Coverage fails for this function (because it isn't being called).  Let's just delete it and pass through an opaque object below for the comparator, e.g.:

```python
inst = self._makeOne(comparator=object())
```

tseaver commented on firestore/tests/unit/v1/test_watch.py:
```diff
@@ -735,6 +735,38 @@ class DummyDoc(object):
+        doc_map = {
+            "/deleted_1": deleted_doc_1,
+            "/deleted_2": deleted_doc_2,
+        }
```
Ugh, `black` wants this reformatted as:

```python
        doc_map = {"/deleted_1": deleted_doc_1, "/deleted_2": deleted_doc_2}
```

liamuk commented:
Changed those two things, and also added you as a collaborator to the patch's fork in case anything else comes up.

You writing code for me to copy into the patch branch isn't the most efficient process haha.

tseaver commented:
The Video Intelligence failure is due to a Kokoro glitch: "Cannot check out non-mergeable pull request as if merged."

tseaver commented:
@liamuk Thanks again for the patch!
